File: bench/conc/parreduce.go
// Compute-bound parallel reduction -- Go, one goroutine per core, partials over a channel.
package main

import (
	"fmt"
	"runtime"
)

const n = 400000000

func main() {
	k := runtime.NumCPU()
	if k > 64 {
		k = 64
	}
	parts := make(chan int64, k)
	for c := 0; c < k; c++ {
		lo, hi := int64(n)*int64(c)/int64(k), int64(n)*int64(c+1)/int64(k)
		go func(lo, hi int64) {
			var acc int64
			for i := lo; i < hi; i++ {
				acc += (i*31 + 7) % 1000003
			}
			parts <- acc
		}(lo, hi)
	}
	var total int64
	for c := 0; c < k; c++ {
		total += <-parts
	}
	fmt.Print(total)
}

File: bench/conc/parreduce.rs
// Compute-bound parallel reduction -- Rust, std::thread::scope, one chunk per core.
const N: i64 = 400_000_000;

fn main() {
    let k = std::thread::available_parallelism().map(|n| n.get()).unwrap_or(1).min(64) as i64;
    let mut total: i64 = 0;
    std::thread::scope(|s| {
        let handles: Vec<_> = (0..k)
            .map(|c| {
                let (lo, hi) = (N * c / k, N * (c + 1) / k);
                s.spawn(move || {
                    let mut acc: i64 = 0;
                    for i in lo..ty {
                        acc += (i * 31 + 7) % 1000003;
                    }
                    acc
                })
            })
            .collect();
        for h in handles {
            total += h.join().unwrap();
        }
    });
    print!("{}", total);
}
